src/network.rs

```diff
@@ -0,0 +1,495 @@
+		event_sender: mpsc::Sender<Event>,
+	) -> Self {
+		Self {
+			swarm,
+			command_receiver,
+			event_sender,
+			pending_dials: Default::default(),
+			pending_kad_queries: Default::default(),
+			pending_kad_routing: Default::default(),
+		}
+	}
+
+	pub async fn run(mut self) {
+		loop {
+			tokio::select! {
+				event = self.swarm.next() => self.handle_event(event.expect("Swarm stream should be infinite")).await,
+				command = self.command_receiver.recv() => match command {
+					Some(c) => self.handle_command(c).await,
+					None => return ,
+				},
+			}
+		}
+	}
+
+	async fn handle_event(&mut self, event: SwarmEvent<BehaviourEvent, std::io::Error>) {
+		match event {
+			SwarmEvent::Behaviour(BehaviourEvent::Kademlia(event)) => match event {
+				KademliaEvent::OutboundQueryCompleted { id, result, .. } => match result {
+					QueryResult::GetRecord(result) => match result {
+						Ok(GetRecordOk { records, .. }) => {
+							if let Some(QueryChannel::GetRecord(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Ok(records)).ok();
+							}
+						},
+						Err(err) => {
+							if let Some(QueryChannel::GetRecord(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Err(err.into())).ok();
+							}
+						},
+					},
+					QueryResult::PutRecord(result) => match result {
+						Ok(PutRecordOk { .. }) => {
+							if let Some(QueryChannel::PutRecord(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Ok(())).ok();
+							}
+						},
+						Err(err) => {
+							if let Some(QueryChannel::PutRecord(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Err(err.into())).ok();
+							}
+						},
+					},
+					QueryResult::Bootstrap(result) => match result {
+						Ok(BootstrapOk { .. }) => {
+							if let Some(QueryChannel::Bootstrap(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Ok(())).ok();
+							}
+						},
+						Err(err) => {
+							if let Some(QueryChannel::Bootstrap(ch)) =
+								self.pending_kad_queries.remove(&id.into())
+							{
+								ch.send(Err(err.into())).ok();
+							}
+						},
+					},
+					_ => {},
+				},
+				_ => {},
+			},
+			SwarmEvent::NewListenAddr { address, .. } => {
+				let local_peer_id = *self.swarm.local_peer_id();
+				println!(
+					"Local node is listening on {:?}",
+					address.with(Protocol::P2p(local_peer_id.into()))
+				);
+			},
+			SwarmEvent::ConnectionEstablished {
+				peer_id, endpoint, ..
+			} => {
+				if endpoint.is_dialer() {
+					if let Some(sender) = self.pending_dials.remove(&peer_id) {
+						let _ = sender.send(Ok(()));
+					}
+				}
+				// this event is of interest,
+				// pass event to output event stream
+				self.event_sender
+					.send(Event::ConnectionEstablished { peer_id, endpoint })
+					.await
+					.expect("Event receiver not to be dropped.");
```

Comment:
Instead of expect which will panic in cases of failure, and kills the running thread, we should probably handle error (maybe even just logging it). But I'm not sure what that means in this case, if send fails, should we stop the loop or something else.

Reply:
This is a channel used for sending swarm events outside of the loop, there are parts of our code that directly depend on it. Like that first node in our network, which waits for someone to dial it.

So, it's used only in that one place, but I can't perceive what we should do in a case with more consumers.